### PR TITLE
implement: validate_environment_marker

### LINKED ISSUE #213: Validate parentheses properly in environment markers
According to [the grammar in PEP 508](https://www.python.org/dev/peps/pep-0508/#complete-grammar), environment markers can have parentheses, like this:

```
foo; extra == "test" and (os_name == "nt" or python_version == "2.7")
```

For now, I just strip leading/trailing parentheses from each clause before checking them. It would be good to check that the parentheses are balanced and only occur in the correct places, before and after individual boolean clauses.

I suspect that this requires a degree of 'proper' parsing of these expressions, rather than just regex-munging.

### PR DESCRIPTION
**Pull request by DreamN**
resolve #213

### CODE REVIEW

File: flit/validate.py [hunk 1]
```diff
@@ -149,37 +153,88 @@ def _valid_version_specifier(s):
...
+                        l, op, r = token[idx - 2:idx + 1]
+                        verify_statement(l, op, r)
+                        stk.append(token[idx])
+                        reduce_expression()
+                        if len(stk) > 1 and stk[-1] in {"and", "or"}:
+                            reduce_expression()
+                    else:
+                        stk.append(token[idx])
+                else:
+                    stk.append(token[idx])
```
Comment (takluyver) [inline flit/validate.py]: Can you add some comments in your parser code? I'm finding it a bit hard to follow.

File: flit/validate.py [hunk 2]
```diff
@@ -149,37 +153,88 @@ def _valid_version_specifier(s):
...
+            idx += 1
+        if len(stk) != 1 or stk[-1] != "EXP":
+            problems.append("Invalid environment markers syntax")
+    except Exception as e:
```
Comment (takluyver) [inline flit/validate.py]: I don't like this structure - it could easily obscure a bug in the code by catching an overly general exception, and it means all the interesting code is indented a level more than it needs to be.

If you want to jump out of parsing as soon as a problem is detected, I think it's OK to use an early return.

File: tests/test_validate.py [hunk 3]
```diff
@@ -76,6 +80,14 @@ def test_validate_environment_marker():
     assert len(res) == 1
     assert res[0].startswith("Invalid expression")
 
+    res = vem("""()))))()extra == "test"(((((((""")  # No chained comparisons
```
Comment (takluyver) [inline tests/test_validate.py]: The comment here is misleading, I think

File: tests/test_validate.py [hunk 4]
```diff
@@ -76,6 +80,14 @@ def test_validate_environment_marker():
...
+    assert len(res) == 1
+    assert res[0] == 'Invalid expression. incorrect parentheses'
+
+    res = vem("""extra == "test" and or (os_name == "nt" or python_version == "2.7")""")  # No chained comparisons
+    assert len(res) == 1
+    assert res[0] == "Invalid expression \"and or\""
+
```
Comment (takluyver) [inline tests/test_validate.py]: Can we add  a test for an empty pair of parentheses, like `extra == "test" and ()`?

File: flit/validate.py [hunk 5]
```diff
@@ -125,17 +127,19 @@ def _is_identifier_attr(s):
...
 VERSION_SPEC = re.compile(r'(~=|===?|!=|<=?|>=?)\s*[A-Z0-9\-_.*+!]+$', re.IGNORECASE)
 REQUIREMENT = re.compile(NAME.pattern[:-1] +  # Trim '$'
-     r"""\s*(?P<extras>\[.*\])?
+                         r"""\s*(?P<extras>\[.*\])?
          \s*(?P<version>[(=~<>!][^;]*)?
          \s*(?P<envmark>;.*)?
      $""", re.IGNORECASE | re.VERBOSE)
 MARKER_OP = re.compile(r'(~=|===?|!=|<=?|>=?|\s+in\s+|\s+not in\s+)')
 
+
```
Comment (takluyver) [inline flit/validate.py]: Not a big deal, but as a general rule, please try not to reformat code beyond the bit that you're working on.

File: flit/validate.py [hunk 6]
```diff
@@ -125,17 +127,19 @@ def _is_identifier_attr(s):
...
     return problems
 
+
 # Distribution name, not quite the same as a Python identifier
 NAME = re.compile(r'^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$', re.IGNORECASE)
 r''
 VERSION_SPEC = re.compile(r'(~=|===?|!=|<=?|>=?)\s*[A-Z0-9\-_.*+!]+$', re.IGNORECASE)
 REQUIREMENT = re.compile(NAME.pattern[:-1] +  # Trim '$'
-     r"""\s*(?P<extras>\[.*\])?
+                         r"""\s*(?P<extras>\[.*\])?
```
Comment (takluyver) [inline flit/validate.py]: Please put the alignment here back how it was, it's like that deliberately.

**Comment by DreamN** [comment]
I'll close this PR and use this PR #225  instead
